examples/input/gamepad_input_events.rs

```diff
@@ -12,27 +12,24 @@ fn main() {
         .run();
 }
 
-fn gamepad_events(mut gamepad_event: EventReader<GamepadEvent>) {
-    for event in gamepad_event.iter() {
-        match event.event_type {
-            GamepadEventType::Connected(_) => {
-                info!("{:?} Connected", event.gamepad);
-            }
-            GamepadEventType::Disconnected => {
-                info!("{:?} Disconnected", event.gamepad);
-            }
-            GamepadEventType::ButtonChanged(button_type, value) => {
-                info!(
-                    "{:?} of {:?} is changed to {}",
-                    button_type, event.gamepad, value
-                );
-            }
-            GamepadEventType::AxisChanged(axis_type, value) => {
-                info!(
-                    "{:?} of {:?} is changed to {}",
-                    axis_type, event.gamepad, value
-                );
-            }
-        }
+fn gamepad_events(
```

Comment:
I'm a little worried that the loss of relative event ordering isn't worth the "improved ergonomics" here.
We already have opinionated abstractions like `Input<GamepadButton>` and `Axis<GamepadAxis>` for users that need fast / ergonomic access to gamepad state for the current frame and aren't worried about relative event ordering. These cover the majority of cases.

GamepadEvent exists for the (much more niche / low level) case of needing less opinionated / frame-independent access to gamepad input. We want it to be as unopinionated as possible because we want to enable as many niche use cases as possible.

Consider the following case: Someone is building a fighting game where input order matters a lot (ex: within a 500ms window, did someone press the "left analog" or "a button" first ... this affects game outcomes ). Their game has pretty graphics and they've noticed that on lower end hardware sometimes they get lag spikes of ~500ms. They don't want the game experience to completely break down when this happens. In this case, using `Input<GamepadButton>` and `Axis<GamepadAxis>` projects the "left analog" and "a button" events onto the "same frame's state", which erases the relevant information. So they choose to write their own input management system to determine what was pressed first.

On Bevy `main`, `GamepadEvent` solves that problem. In this PR, that information is completely lost. They can't even use the `Gilrs` resource directly because we've already consumed the events.

Reply:
Just read this issue / comment: https://github.com/bevyengine/bevy/issues/5984#issuecomment-1345305798

Sounds like you're already thinking about this problem. A global cross event `ts` counter could work. With the t-based approach, we'd also need to make sure we only allow "ordered event reader merging" for events that come from the same source event feed, as the `ts` ordering only makes sense relative to things from the same feed. We could use a global cross-event `ts` counter, but also has problems as it would be "time of processing", not "time of occurrence". It feels like we'd be reinventing the "event ordering wheel" when enums already solve this problem in a simple (and statically checked) way.

Reply:
I agree with this critique: thanks for spotting this Cart. I can't quite figure out how to recreate this ordering in the new version of this PR: can you add or extend an example?

Reply:
You can read the ordered event stream in the same way you previously could.

```rust
fn read_gamepad_events(mut gamepad_events: EventReader<GamepadEvent>) {
   for gamepad_event in gamepad_events.iter() {
       match gamepad_event {
            GamepadEvent::Connection(connection_event) => ()
            GamepadEvent::Button(button_event) => (),
            GamepadEvent::Axis(axis_event) => (),
        }
   }
}
```

But you can also read them individually.
```rust
fn gamepad_axis_event_system(mut axis_events: EventReader<GamepadAxisChangedEvent>) {
    for axis_event in axis_events.iter() {
        // ...
    }
}
```

The event stream is split _after_ `GamepadEvent`s are created, so the order is maintained.

The example demonstrates the latter as that is the preferred way to read the events, whereas the former is for the more advanced use cases where you require the ordered event stream. I've noted in the doc comment of the `GamepadEvent` this it should be read (over the individual events) when in-frame relative ordering is required.

Reply:
Great, thanks for updating the docs and example to more clearly explain this breakdown!